server/src/main/java/io/unitycatalog/server/persist/utils/UriUtils.java

```diff
@@ -0,0 +1,245 @@
+import io.unitycatalog.server.model.AwsCredentials;
+import io.unitycatalog.server.model.AzureUserDelegationSAS;
+import io.unitycatalog.server.model.GcpOauthToken;
+import io.unitycatalog.server.model.GenerateTemporaryModelVersionCredentialsResponse;
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.FileVisitOption;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class UriUtils {
+  private static final Logger LOGGER = LoggerFactory.getLogger(UriUtils.class);
+  private static final ServerPropertiesUtils properties = ServerPropertiesUtils.getInstance();
+
+  private enum Operation {
+    CREATE,
+    DELETE
+  }
+
+  private UriUtils() {}
+
+  // Model specific storage root handlers and convenience methods
+  private static String getModelStorageRoot() {
+    String modelStorageRoot = properties.getProperty("storage-root.models");
+    if (modelStorageRoot == null) {
+      // If the model storage root is empty, use the CWD
+      modelStorageRoot = System.getProperty("user.dir");
+    }
+    // If the model storage root is not a valid URI, make it one
+    if (!isValidURI(modelStorageRoot)) {
+      // Convert to an absolute path
+      File file = new File(modelStorageRoot);
+      modelStorageRoot = file.toURI().toString();
+    }
+    // Check if the modelStorageRoot ends with a slash and remove it if it does
+    while (modelStorageRoot.endsWith("/")) {
+      modelStorageRoot = modelStorageRoot.substring(0, modelStorageRoot.length() - 1);
+    }
+
+    return modelStorageRoot;
+  }
+
+  private static String getModelDirectoryURI(String entityFullName) {
+    return getModelStorageRoot() + "/" + entityFullName.replace(".", "/");
+  }
+
+  public static String getModelStorageLocation(String catalogId, String schemaId, String modelId) {
+    return getModelDirectoryURI(catalogId + "." + schemaId + ".models." + modelId);
+  }
+
+  public static String getModelVersionStorageLocation(
+      String catalogId, String schemaId, String modelId, String versionId) {
+    return getModelDirectoryURI(
+        catalogId + "." + schemaId + ".models." + modelId + ".versions." + versionId);
+  }
+
+  public static String createStorageLocationPath(String uri) {
+    return updateDirectoryFromUri(uri, Operation.CREATE, Optional.empty()).toString();
+  }
+
+  public static String deleteStorageLocationPath(String uri) {
+    return updateDirectoryFromUri(uri, Operation.DELETE, Optional.empty()).toString();
+  }
+
+  private static URI updateDirectoryFromUri(
+      String uri,
+      Operation op,
+      Optional<GenerateTemporaryModelVersionCredentialsResponse> credentials) {
+    URI parsedUri = URI.create(uri);
+    validateURI(parsedUri);
```

Comment:
so we validate the storage Root once and then we validate the entire Uri  again (which is mostly storage root appended with entity Id's)?

Reply:
Yep.  The endpoint handlers introduce new levels to the storage root to generate the final container path of the bucket.  It's probably overkill to do this again, but it's a lightweight check and we did mess with the URI a little before we get to it.